### PR TITLE
RAD-151: L2 and L3 catalog and segmentation map schemas

### LINKED ISSUE #377: Create schemas for source_catalog and segmentation_map
_Issue [RAD-151](https://jira.stsci.edu/browse/RAD-151) was created on JIRA by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva):_

We start with visit level catalogs and segmentation maps.

### PR DESCRIPTION
**Pull request by nden**
Resolves [RAD-151](https://jira.stsci.edu/browse/RAD-151)

Closes #377 

This PR includes and expands Brett's code in #374. 
It adds schemas for Level 2 and Level 3 segmentation maps and source catalog.
This may change as metadata gets finalized by RTB.

**Checklist**
- [ ] Schema changes discussed at RAD Review Board meeting
- [ ] Added entry in `CHANGES.rst` under the corresponding subsection
- [ ] Updated relevant roman_datamodels utilities and tests
- [ ] Passed romancal regression testing on Jenkins / PLWishMaster. Link: https://plwishmaster.stsci.edu:8081/job/RT/job/romancal/XXX/

### CODE REVIEW

File: src/rad/resources/manifests/datamodels-1.0.yaml [hunk 1]
```diff
@@ -263,6 +263,26 @@ tags:
...
+  description: |-
+    Photometry and astrometry computed by the Source Catalog Step
+- tag_uri: asdf://stsci.edu/datamodels/roman/tags/segmentation_map-1.0.0
+  schema_uri: asdf://stsci.edu/datamodels/roman/schemas/segmentation_map-1.0.0
+  title: Segmentation map
+  description: |-
+    Segmentation map computed by the Source Catalog Step
+- tag_uri: asdf://stsci.edu/datamodels/roman/tags/mosaic_source_catalog-1.0.0
+  schema_uri: asdf://stsci.edu/datamodels/roman/schemas/mosaic_source_catalog-1.0.0
+  title: Source catalog
```
Comment (PaulHuwe) [inline src/rad/resources/manifests/datamodels-1.0.yaml:278]:
```suggestion
  title: Mosaic source catalog
```

File: src/rad/resources/schemas/segmentation_map-1.0.0.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            description: |
+              WFI optical element used to take the data.
+            $ref: wfi_optical_element-1.0.0
+          coordinates:
+            tag: asdf://stsci.edu/datamodels/roman/tags/coordinates-1.0.0
+          program:
+            tag: asdf://stsci.edu/datamodels/roman/tags/program-1.0.0
+          visit:
+            tag: asdf://stsci.edu/datamodels/roman/tags/visit-1.0.0
+        required: [optical_element, coordinates, program, visit]
```
Comment (schlafly) [inline src/rad/resources/schemas/segmentation_map-1.0.0.yaml:29]: I don't think we need coordinates.  Good to see the optical element.

File: src/rad/resources/schemas/source_catalog-1.0.0.yaml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            description: |
+              WFI optical element used to take the data.
+            $ref: wfi_optical_element-1.0.0
+          coordinates:
+            tag: asdf://stsci.edu/datamodels/roman/tags/coordinates-1.0.0
+          program:
+            tag: asdf://stsci.edu/datamodels/roman/tags/program-1.0.0
+          visit:
+            tag: asdf://stsci.edu/datamodels/roman/tags/visit-1.0.0
+        required: [optical_element, coordinates, program, visit]
```
Comment (schlafly) [inline src/rad/resources/schemas/source_catalog-1.0.0.yaml:29]: Same comments as for the L2 file; I would drop coordinates, add mid_time_mjd, add effective_exposure_time.

File: src/rad/resources/schemas/mosaic_segmentation_map-1.0.0.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      - $ref: basic-1.0.0
+      - type: object
+        properties:
+          basic:
+            tag: asdf://stsci.edu/datamodels/roman/tags/mosaic_basic-1.0.0
+          coordinates:
+            tag: asdf://stsci.edu/datamodels/roman/tags/coordinates-1.0.0
+          program:
+            tag: asdf://stsci.edu/datamodels/roman/tags/program-1.0.0
+        required: [basic, coordinates, program]
```
[the remark this answers is not in the document]
Comment (nden) [reply src/rad/resources/schemas/mosaic_segmentation_map-1.0.0.yaml]: removed coordinates

File: src/rad/resources/schemas/source_catalog-1.0.0.yaml (hunk 3 above)
Comment (nden) [reply src/rad/resources/schemas/source_catalog-1.0.0.yaml:29]: I removed coordinates. I need to think of the best way to include the two time attributed and then I'll add them.

File: src/rad/resources/schemas/source_catalog-1.0.0.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
+%YAML 1.1
+---
+$schema: asdf://stsci.edu/datamodels/roman/schemas/rad_schema-1.0.0
+id: asdf://stsci.edu/datamodels/roman/schemas/source_catalog-1.0.0
+
+title: Source catalog generated from a Level 3 file by the Source Catalog Step.
```
[the remark this answers is not in the document]
Comment (nden) [reply src/rad/resources/schemas/source_catalog-1.0.0.yaml:6]: Fixed. Thanks!

**Comment by nden** [comment]
I need to think of the best way to include the two time attributed to the level 2 catalogs and then I'll add them (one option is to add the entire exposure schema, though an overkill at the moment, it may end up being simpler after the L1/L2 metadata review). But meanwhile, since all other comments were addressed, I'll open it for a review, particularly to the SDP team, if anything needs to be added to the `archive_catalog` field in those schemas.

File: src/rad/resources/schemas/mosaic_segmentation_map-1.0.0.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      - $ref: basic-1.0.0
+      - type: object
+        properties:
+          basic:
+            tag: asdf://stsci.edu/datamodels/roman/tags/mosaic_basic-1.0.0
+          program:
+            title: Program Information
+            tag: asdf://stsci.edu/datamodels/roman/tags/program-1.0.0
+        required: [basic, program]
+  data:
```
Comment (jbrookens) [inline src/rad/resources/schemas/mosaic_segmentation_map-1.0.0.yaml:24]: Is "data" sufficient for the keyword?

If so, the archive will need to expand on that. The archive_destination should probably be "WFIMosaic.segmentation_map_data".

File: src/rad/resources/schemas/mosaic_source_catalog-1.0.0.yaml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        properties:
+          basic:
+            tag: asdf://stsci.edu/datamodels/roman/tags/mosaic_basic-1.0.0
+          photometry:
+            tag: asdf://stsci.edu/datamodels/roman/tags/photometry-1.0.0
+          program:
+            title: Program Information
+            tag: asdf://stsci.edu/datamodels/roman/tags/program-1.0.0
+        required: [basic, photometry, program]
+  source_catalog:
```
Comment (jbrookens) [inline src/rad/resources/schemas/mosaic_source_catalog-1.0.0.yaml:26]: The archive_destination should probably be "WFIMosaic.source_catalog".

What is the datatype?

File: src/rad/resources/schemas/segmentation_map-1.0.0.yaml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+              WFI optical element used to take the data.
+            $ref: wfi_optical_element-1.0.0
+          program:
+            title: Program Information
+            tag: asdf://stsci.edu/datamodels/roman/tags/program-1.0.0
+          visit:
+            title: Visit Information
+            tag: asdf://stsci.edu/datamodels/roman/tags/visit-1.0.0
+        required: [optical_element, program, visit]
+  data:
```
Comment (jbrookens) [inline src/rad/resources/schemas/segmentation_map-1.0.0.yaml:30]: Is "data" sufficient for the keyword?

If so, the archive will need to expand on that. The archive_destination should probably be "WFIExposure.segmentation_map_data".

File: src/rad/resources/schemas/mosaic_segmentation_map-1.0.0.yaml (hunk 6 above)
Comment (nden) [reply src/rad/resources/schemas/mosaic_segmentation_map-1.0.0.yaml:24]: I would assume you don't want to include `sdf` information about the data array, only the metadata. I thought we may need to add e.g. `WFIMosaicSourceCatalog` to the destination of all metadata keywords. But if that's not the case I think we can merge as is. Let me know if I'm wrong.

Comment (schlafly) [reply src/rad/resources/schemas/mosaic_segmentation_map-1.0.0.yaml:24]: I agree, this is the big data array that you don't want to archive.

File: src/rad/resources/schemas/mosaic_source_catalog-1.0.0.yaml (hunk 7 above)
Comment (schlafly) [reply src/rad/resources/schemas/mosaic_source_catalog-1.0.0.yaml:26]: Likewise, this is the big data array that you don't want to ingest into the archive.

File: src/rad/resources/schemas/segmentation_map-1.0.0.yaml (hunk 8 above)
Comment (schlafly) [reply src/rad/resources/schemas/segmentation_map-1.0.0.yaml:30]: Likewise, this is the big data array that you won't want to ingest into the archive.

File: src/rad/resources/schemas/source_catalog-1.0.0.yaml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+            title: Exposure Information
+            tag: asdf://stsci.edu/datamodels/roman/tags/exposure-1.0.0
+          program:
+            title: Program Information
+            tag: asdf://stsci.edu/datamodels/roman/tags/program-1.0.0
+          visit:
+            title: Visit Information
+            tag: asdf://stsci.edu/datamodels/roman/tags/visit-1.0.0
+        required: [optical_element, exposure, program, visit]
+  source_catalog:
```
[the remark this answers is not in the document]
Comment (schlafly) [reply src/rad/resources/schemas/source_catalog-1.0.0.yaml:33]: Likewise, this is the big data array that you won't want to ingest.

**Comment by jbrookens** [comment]
@nden regarding the archive destinations, please follow this design:

For all keywords in basic, program, and mosaic_basic, add 2 new tables “SourceCatalog” and “SegmentationMap” to the destination (with the same field name as the others).
For all keywords in photometry, add a new table “SourceCatalog” to the destination.
 
And if L2 files will also generate source catalogs and segmentation maps:
 
For all keywords in visit, add 2 new tables “SourceCatalog” and “SegmentationMap” to the destination.
For all keywords in exposure, add a new table “SourceCatalog” to the destination.

**Comment by nden** [comment]
@jbrookens Sounds good. Thanks for the fast decision!

File: src/rad/resources/schemas/source_catalog-1.0.0.yaml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+            $ref: wfi_optical_element-1.0.0
+          exposure:
+            title: Exposure Information
+            tag: asdf://stsci.edu/datamodels/roman/tags/exposure-1.0.0
+          program:
+            title: Program Information
+            tag: asdf://stsci.edu/datamodels/roman/tags/program-1.0.0
+          visit:
+            title: Visit Information
+            tag: asdf://stsci.edu/datamodels/roman/tags/visit-1.0.0
```
Comment (kdupriestsci) [inline src/rad/resources/schemas/source_catalog-1.0.0.yaml:31]: I noticed you didn't make any updates to visit-1.0.0.yaml  Does that mean it should be removed from this file?

**Comment by nden** [comment]
@kdupriestsci @jbrookens Sorry, my previous commit was incomplete. I now got confirmation we will need L2 catalogs and I'll make all changes shortly.

**Comment by nden** [comment]
@PaulHuwe @kdupriestsci @jbrookens This is ready now for a final review.

**Review by kdupriestsci — APPROVED** [review]

File: src/rad/resources/schemas/exposure-1.0.0.yaml
Comment (jbrookens) [inline src/rad/resources/schemas/exposure-1.0.0.yaml:1]: I believe exposure should just be SourceCatalog, not SegmentationMap. My suggestion for L2 was:

And if L2 files will also generate source catalogs and segmentation maps:
 
For all keywords in visit, add 2 new tables “SourceCatalog” and “SegmentationMap” to the destination.
For all keywords in exposure, add a new table “SourceCatalog” to the destination.

Comment (nden) [reply src/rad/resources/schemas/exposure-1.0.0.yaml:1]: Aargh, my bad. Fixed now. Thanks.

**Review by PaulHuwe — APPROVED** [review]
LGTM

**Pull request merged**